app/models/gem_content_entry.rb

```diff
@@ -0,0 +1,122 @@
+class GemContentEntry
+  class InvalidMetadata < RuntimeError; end
+
+  SIZE_LIMIT = 500.megabyte
+  MIME_TEXTUAL_SUBTYPES = %w[json ld+json x-csh x-sh x-httpd-php xhtml+xml xml].freeze
+
+  class << self
+    def from_tar_entry(entry)
+      attrs = {
+        size: entry.size,
+        path: entry.full_name,
+        file_mode: entry.header.mode.to_fs(8)
+      }
+
+      if entry.symlink?
+        new(linkname: entry.header.linkname, **attrs)
+      elsif entry.size > SIZE_LIMIT
+        head = entry.read(4096)
+        mime = Magic.buffer(head, Magic::MIME)
+        new(mime: mime, **attrs)
+      else
+        # read immediately because we're parsing a tar.gz and it shares a single IO across all entries.
+        body = entry.read || ""
+        new(
+          body: body,
+          mime: Magic.buffer(body, Magic::MIME),
+          sha256: Digest::SHA256.hexdigest(body),
+          **attrs
+        )
+      end
+    end
+
+    def from_metadata(metadata, &)
+      attrs = metadata.to_h.symbolize_keys.slice(:body_persisted, :file_mode, :lines, :linkname, :mime, :path, :sha256, :size)
+      raise InvalidMetadata, "missing required keys: #{attrs.inspect}" if attrs[:path].blank? || attrs[:size].blank?
+      attrs[:lines] = attrs[:lines]&.to_i
+      attrs[:body_persisted] = attrs[:body_persisted] == "true"
+      new(persisted: true, **attrs, &)
+    end
+  end
+
+  attr_reader :path, :linkname, :file_mode, :lines, :sha256, :mime, :size
+  alias hex sha256
```

Comment:
Why `hex`? Wouldn't be `hash` better?

Reply:
I went back and forth on this a bunch. I don't know anymore.

It's only internal. It's saved as `sha256` in metadata so we can change hex any time. I don't like hex as much as hash, but hash has special meaning. 🤷

Reply:
[`fingerprint`](https://en.wikipedia.org/wiki/Fingerprint_(computing)) is maybe the most accurate. The only reason I don't use `sha256` everywhere is that I want to have a non algorithm-coupled shorthand that could stay constant no matter how we fingerprint data to identify it. It's a unique identifier for the content separate from the file identity.

Reply:
Yup, clear. Let's use `fingerprint` then.

Reply:
Done.